haystack_experimental/components/tools/tool_invoker.py

```diff
@@ -0,0 +1,204 @@
+            A dictionary with the key `tool_messages` containing a list of ChatMessage objects with tool role
+            that contain the results of the tool invocations.
+
+        :raises ValueError:
+            If the message is not from the assistant role or does not contain tool calls.
+        :raises ToolNotFoundException:
+            If the tool is not found in the list of available tools.
+        :raises ToolInvocationError:
+            If the tool invocation fails.
+        """
+        if not message.is_from(ChatRole.ASSISTANT):
+            raise ValueError("ToolInvoker only supports messages from the assistant role.")
+
+        tool_calls = message.tool_calls
+        if not tool_calls:
+            raise ValueError("ToolInvoker only supports messages with tool calls.")
+
+        tool_messages = []
+
+        for tool_call in tool_calls:
+            tool_name = tool_call.tool_name
+            tool_arguments = tool_call.arguments
+
+            if not tool_name in self._tools_with_names:
+                msg = _TOOL_NOT_FOUND.format(tool_name=tool_name, available_tools=self._tools_with_names.keys())
+                if self.raise_on_failure:
+                    raise ToolNotFoundException(msg)
+                tool_messages.append(ChatMessage.from_tool(tool_result=msg, origin=tool_call))
+                continue
+
+            tool_to_invoke = self._tools_with_names[tool_name]
+            try:
+                tool_response = tool_to_invoke.invoke(**tool_arguments)
+            except ToolInvocationError as e:
+                if self.raise_on_failure:
+                    raise e
+                msg = _TOOL_INVOCATION_FAILURE.format(error=e)
+                tool_messages.append(ChatMessage.from_tool(tool_result=msg, origin=tool_call))
+                continue
+
+            tool_result_string = self._convert_tool_result_to_string(tool_response)
+            tool_messages.append(ChatMessage.from_tool(tool_result=tool_result_string, origin=tool_call))
+
+        return {"tool_messages": tool_messages}
+
+    def to_dict(self) -> Dict[str, Any]:
+        """
+        Serializes the component to a dictionary.
+
+        :returns:
+            Dictionary with serialized data.
+        """
+        serialized_tools = [tool.to_dict() for tool in self.tools]
+        return default_to_dict(self, tools=serialized_tools, raise_on_failure=self.raise_on_failure)
+
+    @classmethod
+    def from_dict(cls, data: Dict[str, Any]) -> "ToolInvoker":
+        """
+        Deserializes the component from a dictionary.
+
+        :param data:
+            The dictionary to deserialize from.
+        :returns:
+            The deserialized component.
+        """
+        init_params = data.get("init_parameters", {})
+        serialized_tools = init_params.get("tools", [])
+        deserialized_tools = [Tool.from_dict(tool_data) for tool_data in serialized_tools]
+        data["init_parameters"]["tools"] = deserialized_tools
```

Comment:
Let's add a utility function similar to `deserialize_secrets_inplace` for tools? I think we'll need it once we start supporting more generators.

Reply:
I created a separated PR for that: #61
Once the PR is merged, I will rebase this PR and use the utility function

Reply:
done